### ISSUE TITLE
Difference between the code and Eq(13) in the paper about the gradient calculation

### ISSUE THREAD

**Issue by dywu98** [Gsunshine/Enjoy-Hamburger#9]
https://github.com/Gsunshine/Enjoy-Hamburger/blob/d9b51f6f197486df68c6e059e396520680157c08/seg_mm/mmseg/models/decode_heads/ham_head.py#L45

According to the paper, isn't the gradient of the MDs should be one-step gradient? However, the code of NMF dose not apply `torch.with_on_grad()` on the `local_inference` of NMF and `_MatrixDecomposition2DBase`. Could you please provide some explanation on this difference?

**Comment by Gsunshine** [comment]
Hi @Magic-Ha ,

Thank you for your interest in Hamburger! This is an excellent question! 

The gradient steps you can use depend on the conditioning of the Jacobian matrix. For NMF in LightHam, according to my test, the 1-step gradient produces similar performances as the exact gradient. So you can definitely choose 1-step gradient for NMF to save your backdrop memory and time (will be both O(1) memory and time complexity in the backward pass). (Remove the # there.)
For the optimization algorithm in VQ, however, the conditioning will usually prevent you from exploiting more gradient information there. So if you are going to choose the VQ Ham for Hamburger, it is generally better to directly use 1-step grad.

As for further analysis, you may check my paper on training more generalized implicit layers/deep equilibrium models [here](https://arxiv.org/abs/2111.05177). To improve the gradient's conditioning and incorporate more gradient information, you may consider a ``damping'' effect that can form a descent direction compared to the exact gradient (with better conditioning) while keeping your fixed point/optimization solution unchanged.

One more time, the conditioning of Jacobian determines how exact the gradient should be. So it is not strange that inexact gradients like 1-step gradient can produce as good results as the exact gradient or even surpass it sometimes.

Embrace inexact gradients and enjoy Hamburger!
Thank you for your interest again! 

Zhengyang

**Comment by dywu98** [comment]
Thanks for your explanation. You've been very helpful！

**Comment by Gsunshine** [comment]
Please feel free to ask if you have additional questions. I'm very happy to chat and explain. :)

**Comment by dywu98** [comment]
Sorry about the typo lol. Maybe I click the wrong prompt. I do mean to type explanation :) . Already corrected it.

**Comment by Gsunshine** [comment]
Will be an interesting topic for the NLP conference, lol. 
Investigating and evaluating the prompt bias in pretrained language models for typewriting software. EMNLP 2023
